Recoloured and added waves

Adubbz opened the pull request:
This PR makes a few aesthetic improvements to the menu. Namely:

- Added waves to the bottom of the screen (and re positioned elements accordingly)
- Changed the background and text colour
-  Made the highlight around the active entry thicker

The new design can be seen in action [here](https://gfycat.com/KeenOrderlyBordercollie)

And on actual hardware [here](https://www.youtube.com/watch?v=a4liMi8xLXY) (courtesy of VAL-001)

An alternate mode is also included, however it must be manually changed and compiled. It can be seen [here](https://gfycat.com/UnfortunateSkeletalHound)

Adubbz commented:
There was an issue with a single line of pixels not filled in in the above images, that has been addressed now.

plutooo commented:
https://i.imgur.com/RkPrUYK.jpg

Make it look like this? :)

jaames commented:
Plutoo mentioned that the intention was for hbl to match the switch theme, so here's a light variant of the same mockup:

https://i.imgur.com/X6C0SFU.jpg

Tomorrow I'll try to polish both of them a bit further

Adubbz commented:
I updated it to be more in line with the concept art. See [here](https://gfycat.com/MisguidedNaturalCanvasback)

CTCaer commented:
@Adubbz 
Great job!

Can you try the following small aesthetic changes:
1. Remove the Name text border/fill (the rectangle on top of small icons).
2. Name text left aligned to the start of white icon border (like Switch home menu game titles)
3. Name text a little more higher (around the vertical center of the name text border).
4. Dark directory text color. 

It will look like this:
![](https://i.imgur.com/lC0Bjn3.png)

Maufeat commented:
I like how it currently looks. Those changes above aren't looking ordered and clean. Should include "banner.jpg" for the top banner for each hbapp (bigger width than icon) everything else looks perfectly fine imo.

jaames commented:
![image](https://user-images.githubusercontent.com/9112876/36488563-a970135c-171b-11e8-8931-930dda103cce.png)

Since square icons are also being used for the large "details" view, I think it might look better if they're moved to the other side of the app title and description text

MrKcyre commented:
wow nice job guys! this last concept looks perfect! the blue gradient in the wave detail is something awesome! And it looks super clean with the main text on the right side. nice job!

jaames commented:
Thanks @MrKcyre!

Image assets and details on layout, colors, fonts, etc from my mockups can now be found here: https://github.com/jaames/switch-hbl-mockup :)

Also imo the waves are a little tall right now? I intended for the wave effect to be *super* subtle, with an amplitude of ~8px or so at the most, they feel a bit distracting otherwise

yellows8 commented:
Probably should wait to add time-display until local-time is supported, don't think most users want UTC.

fincs commented:
This is kind of a silly nitpick, but there seems to be kind of an identity crisis in this project. The repository name is 'nx-hbmenu' but the program calls itself "The Homebrew Launcher" in current upstream. Having this be abbreviated to "hbl" in the redesign above reminded me of this. Can we do something about this? (Personally I've never liked the "hbl" abbreviation since it makes me think of HBlank - as a result I always call it 'hbmenu').

jaames commented:
@fincs raises a good point imo. While I'm not particularly tied to either name, I thought it might be worth trying a mockup using a "hbmenu" logo for the sake of comparison:

<img width="637" alt="" src="https://user-images.githubusercontent.com/9112876/36513114-e12b4632-1765-11e8-9d50-9d3301771e8e.png">

Adubbz commented:
I changed the name, upped the version number to 2.0.0 and hid the author and version for folders. I think it's good to go now. [Preview](https://gfycat.com/AdorablePettyBluejay)

plutooo commented:
Looks sweet, can we get final comments from James

plutooo commented:
I think I can see a few nitpicks -- missing bold font on the logo, and lack of rounded corners

plutooo commented:
Let's do a full 2.0.0 release when jaames gives the approval. Awesome work so far

Adubbz commented:
Yeah, fonts do indeed need changing but those are waiting on @fincs’s tool to be made public. I personally prefer non-rounded corners, plus they’re more in line with the real home menu. Will see what Jaames thinks

plutooo commented:
I feel like I'm asking for too much, but how about animation when moving sideways across the icons? Right now it's just instant lacking any smooth scrolling afaik?

Adubbz commented:
Yeah it is, I'll see what I can do :)

jaames commented:
Looking good! I do have a couple of nitpicks though:

* The waves are still a bit tall imo, as mentioned earlier. I think the effect should be very subtle so it's almost just a border between the top of the app list and the rest of the UI. In the future maybe the waves could be configured via theme settings for people that want more/no waves?

* I did intend for the app list tiles to be square icons without any text, like the Switch menu, although that does kind of rely on homebrew authors making recognisable icons. As such I'm fine with titles being there too, but to take the extra height into account I'd, again, tone the wave amplitude down and then shift both the wave level and app list upwards by about 10 - 20 pixels to give the directory path text more space.

* The rightmost item in the app list should be allowed to fall off the edge of the screen slightly, to give the user an indication that the list can be scrolled to reveal more offscreen items.

* It would be great if the gradient effect for the bottom ~180px of the waves could be added, it makes the waves look like they have depth and the directory path is a little more readable on the darker blue imo.

* As for rounded tile corners, I really don't mind personally, slight shadows under the icons would look really nice though.

also, if you want to incorporate the bold "hbmenu" logo from the last mockup [there's a transparent PNG version here](https://github.com/jaames/switch-hbl-mockup/blob/master/assets/dark/hbmenuLogoSmall.png) :)

x43x61x69 commented:
Nice! However the margin values and colors could be made into const via `define` to simplify the editing. (Also make the code more understandable.)

I do some quick editing and made one that mimic the official system UI w/ animation:

![](https://thumbs.gfycat.com/PaleLeftIcelandicsheepdog-size_restricted.gif)

Adubbz commented:
See my light theme commit @x43x61x69, all the relevant colour values are now in the theme_t struct which is initialized in themeStartup

jaames commented:
@Adubbz Oh cool, I almost didn't notice because it was so subtle lol

By the way, looking at these previews again, I think my issue with the waves is that the purple background wave layer is too visible? In my mockups the two purple background waves only just occasionally peek up behind the two foreground ones, but if people prefer it that way then I won't complain. 

I have to say, the wave motion itself is looking really good though :)

Adubbz commented:
I'm a bit confused what you want change with respect to the waves... It sounded like you wanted them lowered... then raised again? Is it just the background waves that should be lowered?

gdkchan commented:
I agree that the waves are too tall. Would be better if the distance between the lowest point on the cyan wave, and the peak of the purple wave was reduced. As it is currently its too distracting.

Adubbz commented:
I believe everything that was mentioned has been addressed now. 

[Dark theme](https://gfycat.com/BlankChillyBarasinga)

[Light theme](https://gfycat.com/UncomfortableRectangularAfricanelephant)

fincs commented on common/math.h:
```diff
@@ -0,0 +1,13 @@
+}
+
+inline float nxfabs(float x) {
+    return x < 0 ? -x : x;
```
What's the point of replacing fabs with this function? GCC is able to see fabs calls and replace it with actual AArch64 VFP instructions.

Adubbz replied:
I was having troubles with including math.h in the nro builds when I was trying to use sin previously, so I assumed fabs would be the same. It seems to work fine so I've changed it now. In all likelihood I was just doing something wrong before.

Adubbz commented:
I was having troubles with including math.h in the nro builds when I was trying to use sin previously, so I assumed fabs would be the same. It seems to work fine so I've changed it now

jaames commented:
The latest previews are looking pretty good :)

It would definitely still be nice to have better fonts, custom logo, clock, etc, but I guess they can be left for future releases if it's not possible to implement them at the moment

By the way, would it be worth making new folder icons that match each theme?

yellows8 commented:
setsysExit() missing.

plutooo commented:
The border around the currently selected icon seems to be assymetrical?

plutooo commented:
Also missing the (A) button description in the lower right corner

Maufeat commented:
I would also love to implement the blinking active border (which @x43x61x69 implemented on his theme)
And I can't compile it anymore, someone got the same issue? (most recent libnx git and devkitA64) 
![image](https://user-images.githubusercontent.com/3602392/36562685-05c55cdc-1818-11e8-9ce9-e72041bbc325.png)

yellows8 commented:
The value used for updating "move" in menuUpdate() for KEY_DOWN/KEY_UP should be changed to match {max number of displayed entries}.

Adubbz commented:
I've added the a button description and changed the hbmenu logo to jaames' one.

![image](https://user-images.githubusercontent.com/2765389/36576992-71668fac-18a8-11e8-8262-dfd8cefd06e5.png)

![image](https://user-images.githubusercontent.com/2765389/36576997-79effa28-18a8-11e8-8923-8ecbc9ed5031.png)

Maufeat commented:
Messed up that commit message, but it should now show "Open" instead of "Launch" when a folder is selected.

fincs commented:
I think it might be also a good idea to add "(B) Back" text in non-root folders to indicate that you can go up one directory.

plutooo commented:
Nice, Now fix the font and we're ready to go!

plutooo commented:
Also seems like the dark theme is blue when it should be grey

MrKcyre commented:
@plutooo i remember the default dark theme on TheHomebrewChannel (wii) and HomebrewLauncher (DS and WiiU) were blue but darker as @jaames used this design pattern

jaames commented:
Adding saturation to grey to get a slight blue tint is a common design trick tbh, Discord, Twitter's night mode and Photoshop all do this for example... I probably overdid it slightly :P 

Something like `#2D2D36` might be better for the dark theme background color

Adubbz commented:
![screenshot from 2018-02-24 12-10-45](https://user-images.githubusercontent.com/2765389/36623796-1de596de-195c-11e8-8095-db209e715498.png)
![screenshot from 2018-02-24 12-11-06](https://user-images.githubusercontent.com/2765389/36623798-1f31ecae-195c-11e8-89c1-04edd4f6ce8c.png)

All good? :)

anowlcalledjosh commented:
The Launch action looks like it might be too close to the edge of the screen now, judging by the margin to the left of the hbmenu logo.

jaames commented:
@Adubbz could we maybe shift the app list tiles up by 10 - 20 px or so? The text at the bottom seems a little cramped at the moment so it could use some more space, otherwise it looks great! :>

Adubbz commented:
I moved up the app icons by 5, any more and I think it's too close to the waves, which I don't think should be moved further up either. I also shifted the launch icon to the left a tad more. 

![screenshot from 2018-02-24 12-20-53](https://user-images.githubusercontent.com/2765389/36623923-a38d58ca-195d-11e8-9916-8a4e79f39449.png)
![screenshot from 2018-02-24 12-23-54](https://user-images.githubusercontent.com/2765389/36623924-a3badc5a-195d-11e8-8fdf-4a3262e0a23e.png)

MrKcyre commented:
omgg it is so beautiful! awesome work @Adubbz !

jaames commented:
The text at the bottom still feels a little cramped to me tbh, we can wait and see what others think though \ o /

An option might be to make the app details icon a bit smaller, then the wave level and app list can be shifted upwards just a little more? The details icon looks like it's using the full 256x256 size, while in my mockup I had it sized at 220x220 so I could be more generous with space around the bottom text

Adubbz commented:
Alright, had a chat with jaames and I think it's good to go now. We changed the size and positioning of the bottom text a little bit to give a touch more space than before.

![screenshot from 2018-02-24 13-01-25](https://user-images.githubusercontent.com/2765389/36624423-10e466ca-1963-11e8-8370-43f2fb668415.png)
![screenshot from 2018-02-24 13-02-03](https://user-images.githubusercontent.com/2765389/36624424-11c558f6-1963-11e8-983c-99e11243c7ff.png)

plutooo commented:
Nice. Is the selection border supposed to be purple? Looks a bit off

Also need a variation on the default_icon for invalid-icon-icon, and a new folder icon. Cheers

Maufeat commented:
Someone else has this issue? Most of my entries are out of screen.
![20180224_171149](https://user-images.githubusercontent.com/3602392/36632368-3fdca974-1986-11e8-8fe3-0c6a6a5b01c8.jpg)

yellows8 commented:
DrawTextTruncate() should be used for the path like app-entry-title already is, since it could overlap with {button text} when long enough.

Maufeat commented:
done. But we should probably adding the direction from where to truncate since I would rather truncate the path from left.

jaames commented:
Some suggestions for new default / invalid / folder icons:

<img width="526" alt="" src="https://user-images.githubusercontent.com/9112876/36634292-3866c314-199a-11e8-91d9-0312f6bbc980.png">

I wasn't sure if we wanted to reuse the question block since it's now the default libnx icon

fincs commented:
I like "default_icon alt." and "invalid_icon_icon alt." better than the default libnx icon which was merged recently - they are closer to the original icons and they keep in line with the flat theme hbmenu is supposed to have.

Adubbz commented:
Icons are updated:

![screenshot from 2018-02-25 10-27-36](https://user-images.githubusercontent.com/2765389/36636250-af44df94-1a16-11e8-90fd-6c8d614e2732.png)
![screenshot from 2018-02-25 10-28-09](https://user-images.githubusercontent.com/2765389/36636251-af700e12-1a16-11e8-8436-f830c20a9213.png)

yellows8 commented:
These strings should be loaded via ```textGetString(StrId_{name})```, not hard-coded. However, once get-language is supported, the length of the strings would vary for non-English...

https://github.com/switchbrew/nx-hbmenu/blob/3081701a21cb9c1e577f7413dfa8532bb09a99a5/common/menu.c#L201-L205

yellows8 commented:
README should be updated since the icons were updated.

CuriousTommy commented:
@jaames Would it make sense to have a working application show a switch icon with a question mark? IMO, I feel like you would normally see a question mark where there is doubt (in other words, when something is not right).

Adubbz commented:
There is doubt however, because the application is missing an icon and therefore has reverted to the default

Adubbz commented:
I removed having any credits for the icons in the README, since that's what @jaames did in his PR to libnx

CuriousTommy commented:
True... Personally, I think having a generic switch icon for working applications and a question mark (or exclamation mark) switch for invalid applications is would be better. I feel like a user might see the question mark and might think that there is something wrong with the application.

But regardless, this is just my opinion.

yellows8 commented:
Well, the original libnx icon didn't have any '?'/'!'.

@CuriousTommy The hbmenu default-icon is only shown when loading the icon failed, normally besides that nothing is "invalid" with the application.

Maufeat commented:
Performance is now way better, but anti aliasing is still causing drops. When anti aliasing on waves is disabled it runs at full speed.

Caching vs not caching doesn't make any difference here, probably a LUT would help here but as Adubbz already said there are way to many variants to precalc those.

MrKcyre commented:
@Adubbz just a little detail.. i know the translations won't work right now but.. I'm portuguese and i think the best translation for 'launch' would be 'Executar'.. And the best translation for 'Open' would be 'Abrir' instead of 'Aberto'.. The rest is perfect!

Maufeat commented:
Fixed the AA performance by reverting and removing the caching and instead of checking `if(alpha < 0.3)` I now check `if(alpha >= 0.3)` and put the anti aliasing into the `else`.

And @MrKcyre we should put the translation into another PR - there are lots a few improvements to be made. For example `Version` in Germany is still `Version` and not `Ausführung` (doesn't sound right)

MrKcyre commented:
Oh, alright @Maufeat

Adubbz commented:
I changed the icon/text placement to follow the golden ratio as per naehrwert's suggestion, after he pointed out that it didn't really have any sort of alignment to anything, and I'm inclined to agree. Right now the icon is centred in the shorter section, whilst the text is left-aligned to the larger section. 

In terms of rounded corners, I think they wouldn't work too well with the row of cards on the bottom. They were fine in the original concept because it assumed it was just the app icons showing. However in this case we also have to have the titles, as there will be a few homebrew authors who don't bother to include a distinctive icon. With rounded corners I think they will end up looking pill-shaped, which is pretty inconsistent with all of the other UIs on the switch.

Of course, the other option is to round only the top icon. But again, I think consistency is important. Having some things with rounded corners and others without doesn't seem like a good idea personally.
